Инициативы

nurlantr opened the pull request.

nurlantr committed:
Инициативы

nurlantr committed:
Merge branch 'murloc-master-bc5' into feature-BJS2-15767

# Conflicts:
#	src/main/java/faang/school/projectservice/model/initiative/Initiative.java

raukaztaev commented on src/main/java/faang/school/projectservice/controller/initiative/InitiativeController.java:
```diff
@@ -0,0 +1,55 @@
+public class InitiativeController {
+    private final InitiativeService initiativeService;
+
+    @PostMapping("/create")
```
почитай про именование ресурсов в Rest(https://restapitutorial.ru/lessons/restfulresourcenaming/)
В URI не должны быть глаголы, тип действия определяет метод http(в данном случае POST "/initiavites")
То же самое для эндпоинта ниже: update

nurlantr committed:
в InitiativeController убрал create и update.

nurlantr committed:
Merge branch 'murloc-master-bc5' into feature-BJS2-15767

egorbarzenkov commented on src/main/java/faang/school/projectservice/controller/initiative/InitiativeController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @PostMapping
+    public InitiativeDto createInitiative(@RequestBody InitiativeDto initiativeDto) {
+        if (initiativeDto.getName() == null || initiativeDto.getName().isEmpty() || initiativeDto.getName().isBlank()) {
```
лучше это унести на уровень валидации через аннотации на InitiativeDto, можно пометить поле 
@NotBlank(message = "name is blank") (сообщение об ошибке придумай сам) и с помощью аннотации
@Valid можно уже не валидировать это поле руками
примерно так будет выглядеть параметр в методе
@RequestBody @Valid InitiativeDto initiativeDto

egorbarzenkov commented on src/main/java/faang/school/projectservice/controller/initiative/InitiativeController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PutMapping("/{id}")
+    public InitiativeDto updateInitiative(@PathVariable("id") Long id, @RequestBody InitiativeDto initiativeDto) {
```
у тебя в initiativeDto есть поле private Long id; разве оно не является идентификатором инициативы?
если так, то параметр @PathVariable("id") Long id становится не нужен и даже вреден, может случится так, что придет запрос с разными @PathVariable("id") Long id и initiativeDto.id тогда логически данные будут несогласованны что приведет к некорректному обновлению инициативы

egorbarzenkov replied:
если же на разных этапах работы с InitiativeDto нужна разная валидация то почитай про аннотацию validated
https://www.baeldung.com/spring-valid-vs-validated

egorbarzenkov commented on src/main/java/faang/school/projectservice/controller/initiative/InitiativeController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @GetMapping("/filter")
+    public List<InitiativeDto> getAllInitiativesWithFilter(@RequestParam InitiativeStatus status,
```
лучше 2 этих параметра объединить в один InitiativeFilterDto, чтобы при разрастании условий фильтрации не пухло количество параметров передаваемых в метод и соответственно в path запроса

egorbarzenkov commented on src/main/java/faang/school/projectservice/service/initiative/InitiativeService.java:
```diff
@@ -0,0 +1,97 @@
+
+    public List<InitiativeDto> getAllInitiativesWithFilter(InitiativeStatus status, Long curatorId) {
+        return initiativeRepository.findAll().stream()
+                .filter(initiative -> (status == null || initiative.getStatus() == status))
```
если объем фильтров будет возрастать, то придется лезть и в этот код, лучше переделать фильтры на бины и инжектить их по типу в сервис и здесь уже применять их либо если не планируется расшивание фильтрации, то можно опустить эту фильтрацию на уровень БД и тащить в код сразу все необходимые данные

egorbarzenkov commented on src/main/java/faang/school/projectservice/service/initiative/InitiativeService.java:
```diff
@@ -0,0 +1,97 @@
+    public InitiativeDto getInitiativeById(Long id) {
+        return initiativeRepository.findById(id)
+                .map(initiativeMapper::toDto)
+                .orElseThrow(() -> new RuntimeException("Initiative not found"));
```
отходит от слишком общих ошибок, посмотри в сторону EntityNotFoundException и так же в сообщение закинь идентификатор, по которому не было найдено сущности

egorbarzenkov commented on src/main/java/faang/school/projectservice/service/initiative/InitiativeService.java:
```diff
@@ -0,0 +1,97 @@
+        boolean isEmptyDescription = initiativeDto.getDescription().isEmpty();
+        boolean isBlankDescription = initiativeDto.getDescription().isBlank();
+
+        List<Initiative> initiatives = initiativeRepository.findAll();
```
очень плохой способ валидацииты дважды затягиваешь полную таблицу initiative и уже с ней проводишь манипуляции, подумай в сторону более точечного запроса для проверки информации об инициативе в рамках проекта

nurlantr replied:
как я понял мне нудно в классе InitiativeRepository создать методы с конкретным запросом в бд?
Методы по типу этих:
boolean existsByNameAndProjectId(String name, Long projectId);
boolean existsByProjectIdAndStatus(Long projectId, InitiativeStatus status);

egorbarzenkov replied:
да, ты все верно понял, можно попытаться уместить это в один метод, получения по идентификатору и уже в полученной сущности проверять статус и описание

egorbarzenkov commented on src/main/java/faang/school/projectservice/service/initiative/InitiativeService.java:
```diff
@@ -0,0 +1,97 @@
+            throw new IllegalArgumentException("Description cannot be empty");
+        }
+        if (initiativeExist) {
+            throw new IllegalArgumentException("Initiative with the same name already exists in this project");
```
персонализируй ошибки, чтобы в них была информацию о том, какая запись не прошла валидацию

nurlantr replied:
Мне получается нужно создать свои классы исключений для каждого из них? и все они должны наследовать Runtime Exception?

egorbarzenkov replied:
нет, у тебя не информативны сами сообщения, если ты будешь указывать хотя бы идентификатор по которому идешь в базу, то потом будет проще разбирать что пошло не так

nurlantr committed:
Merge branch 'murloc-master-bc5' into feature-BJS2-15767

nurlantr committed:
Изменения:
1) Обновил эндпоинты и ApiPath
2) Добавил аннотации в класс InitiativeDto и теперь использую @Valid
3) В InitiativeController в методе updateInitiative убрал path variable
4) Создал класс InitiativeFilterDto для метода getAllInitiativesWithFilter
5) Оптимизировал метод validateCreateInitiative в классе InitiativeService

nurlantr committed:
Merge branch 'murloc-master-bc5' into feature-BJS2-15767

egorbarzenkov commented on src/main/java/faang/school/projectservice/service/initiative/InitiativeService.java:
```diff
@@ -0,0 +1,96 @@
+    private final List<InitiativeFilter> filters;
+
+    public InitiativeDto createInitiative(InitiativeDto initiativeDto) {
+        validateCreateInitiative(initiativeDto);
```
не хватает проверки на "У куратора инициативы должна быть соответствующая роль"
так же нет функционала для разделения инициативы на этапы "Инициатива сразу может быть разделена на этапы при создании", для этот нужно расширить DTO и заложить функционал на то, что в пришедшей инициативе могуб быть несуществующие и уже существующие этапы
